### ISSUE TITLE
Async work with input blobs

### ISSUE THREAD

**Issue by SergeyMalashenko** [BVLC/caffe#6556]
Good day!
I want to fill input blobs in async mode. So I need to know when the first layer of  the model has completed to work and I can can continue to work with input blobs. How to do it?

**Comment by csukuangfj** [comment]
Relevant [code][1] is below
```cpp
template <typename Dtype>
Dtype Net<Dtype>::ForwardFromTo(int start, int end) {
  CHECK_GE(start, 0);
  CHECK_LT(end, layers_.size());
  Dtype loss = 0;
  for (int i = start; i <= end; ++i) {
    for (int c = 0; c < before_forward_.size(); ++c) {
      before_forward_[c]->run(i);
    }
    Dtype layer_loss = layers_[i]->Forward(bottom_vecs_[i], top_vecs_[i]);
    loss += layer_loss;
    if (debug_info_) { ForwardDebugInfo(i); }
    for (int c = 0; c < after_forward_.size(); ++c) {
      after_forward_[c]->run(i);
    }
  }
  return loss;
}
```

You can put some callbacks using [add_after_forward()][2].

The index of the input layer is 0, so if your callback receives an argument 0, then it means the input layer has finished its forward pass.

Be aware that if the input blobs contain labels, you have to wait until the last layer finishes computing the loss.

Caffe provides `DataLayer` to load input in another thread using a blocking queue.

[2]: https://github.com/BVLC/caffe/blob/99bd99795dcdf0b1d3086a8d67ab1782a8a08383/include/caffe/net.hpp#L243
[1]: https://github.com/BVLC/caffe/blob/99bd99795dcdf0b1d3086a8d67ab1782a8a08383/src/caffe/net.cpp#L518

**Comment by SergeyMalashenko** [comment]
Thank you!
